packages/query/src/helpers/identify-txs.ts

```diff
@@ -0,0 +1,137 @@
+  }
+
+  return tx.body.actions
+    .flatMap(({ action }) => {
+      switch (action.case) {
+        case 'output':
+          return action.value.body?.notePayload?.noteCommitment;
+        case 'swap':
+          return action.value.body?.payload?.commitment;
+        case 'swapClaim':
+          return [action.value.body?.output1Commitment, action.value.body?.output2Commitment];
+        default:
+          return;
+      }
+    })
+    .filter(isDefined);
+};
+
+export const getNullifiersFromActions = (tx: Transaction): Nullifier[] => {
+  if (!tx.body?.actions) {
+    return [];
+  }
+
+  return tx.body.actions
+    .flatMap(({ action }) => {
+      switch (action.case) {
+        case 'spend':
+        case 'swapClaim':
+          return action.value.body?.nullifier;
+        default:
+          return;
+      }
+    })
+    .filter(isDefined);
+};
+
+export interface RelevantTx {
+  id: TransactionId;
+  data: Transaction;
+}
+
+type RecoveredSourceRecords = (SpendableNoteRecord | SwapRecord)[];
+
+const generateTxId = async (tx: Transaction): Promise<TransactionId> => {
+  return new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
+};
+
+const searchRelevant = async (
+  tx: Transaction,
+  spentNullifiers: Set<Nullifier>,
+  commitmentRecords: Map<StateCommitment, SpendableNoteRecord | SwapRecord>,
+): Promise<
+  { relevantTx: RelevantTx; recoveredSourceRecords: RecoveredSourceRecords } | undefined
+> => {
+  let txId: TransactionId | undefined; // If set, that means this tx is relevant and should be returned to the caller
+  const recoveredSourceRecords: RecoveredSourceRecords = [];
+
+  const txNullifiers = getNullifiersFromActions(tx);
+  for (const spentNullifier of spentNullifiers) {
+    if (txNullifiers.some(txNullifier => spentNullifier.equals(txNullifier))) {
+      txId ??= await generateTxId(tx);
+    }
+  }
+
+  const txCommitments = getCommitmentsFromActions(tx);
+  for (const [stateCommitment, spendableNoteRecord] of commitmentRecords) {
+    if (txCommitments.some(txCommitment => stateCommitment.equals(txCommitment))) {
+      txId ??= await generateTxId(tx);
+
+      // Blank sources can be recovered by associating them with the transaction
+      if (BLANK_TX_SOURCE.equals(spendableNoteRecord.source)) {
+        const recovered = spendableNoteRecord.clone();
+        recovered.source = new CommitmentSource({
+          source: { case: 'transaction', value: { id: txId.inner } },
+        });
+        recoveredSourceRecords.push(recovered);
+      }
+    }
+  }
+
+  if (txId) {
+    return {
+      relevantTx: { id: txId, data: tx },
+      recoveredSourceRecords,
+    };
+  }
+
+  return undefined;
+};
+
+// identify transactions that involve a new record by comparing nullifiers and state commitments
+// also returns records with recovered sources
+export const identifyTransactions = async (
+  spentNullifiers: Set<Nullifier>,
+  commitmentRecords: Map<StateCommitment, SpendableNoteRecord | SwapRecord>,
```

Comment:
think we might have another potential bug. I don’t think we've ever addressed the case where the map holding `StateCommitment` objects contains [duplicate commitments](https://github.com/penumbra-zone/penumbra/issues/3945)?

Reply:
In that case, the worst that could happen is in this line: https://github.com/penumbra-zone/web/pull/1793/files#diff-df9a3308a5ae26e8f5e912c4bdd659a76d12e6f751ec7c3eb8de3c0b81147bf0R91-R98 where there are two duplicate recovered records. And in that case, we just [save to the database](https://github.com/penumbra-zone/web/blob/75ef4f5bb7bb07c4ea0478bee360beba12220f28/packages/query/src/block-processor.ts#L424-L440) twice. Don't see a failure case we need to worry about.